tests/helpers/test_script.py

```diff
@@ -315,837 +243,511 @@ def heard_event_cb(event):
+@pytest.mark.parametrize("count", [1, 3])
+@pytest.mark.parametrize("script_mode", _BASIC_SCRIPT_MODES)
+async def test_stop_no_wait(hass, caplog, script_mode, count):
+    """Test stopping script."""
+    service_started_sem = asyncio.Semaphore(0)
+    finish_service_event = asyncio.Event()
+    event = "test_event"
+    events = async_capture_events(hass, event)
 
-    delay = timedelta(milliseconds=10)
-    schema = cv.SCRIPT_SCHEMA({"delay": delay, "alias": delay_alias})
+    async def async_simulate_long_service(service):
+        """Simulate a service that takes a not insignificant time."""
+        service_started_sem.release()
+        await finish_service_event.wait()
 
-    for run_mode in _ALL_RUN_MODES:
-        delay_started_flag.clear()
+    hass.services.async_register("test", "script", async_simulate_long_service)
 
-        if run_mode is None:
-            script_obj = script.Script(hass, schema, change_listener=delay_started_cb)
-        else:
-            script_obj = script.Script(
-                hass, schema, change_listener=delay_started_cb, run_mode=run_mode
-            )
+    sequence = cv.SCRIPT_SCHEMA([{"service": "test.script"}, {"event": event}])
+    script_obj = script.Script(hass, sequence, script_mode=script_mode)
 
-        assert script_obj.can_cancel
+    # Get script started specified number of times and wait until the test.script
+    # service has started for each run.
+    tasks = []
+    for _ in range(count):
+        hass.async_create_task(script_obj.async_run())
+        tasks.append(hass.async_create_task(service_started_sem.acquire()))
+    await asyncio.wait_for(asyncio.gather(*tasks), 1)
 
-        try:
-            if run_mode == "background":
-                await script_obj.async_run()
-            else:
-                hass.async_create_task(script_obj.async_run())
-            await asyncio.wait_for(delay_started_flag.wait(), 1)
+    # Can't assert just yet because we haven't verified stopping works yet.
+    # If assert fails we can hang test if async_stop doesn't work.
+    script_was_runing = script_obj.is_running
+    were_no_events = len(events) == 0
 
-            assert script_obj.is_running
-            assert script_obj.last_action == delay_alias
-        except (AssertionError, asyncio.TimeoutError):
-            await script_obj.async_stop()
-            raise
-        else:
-            if run_mode in (None, "legacy"):
-                future = dt_util.utcnow() + delay
-                async_fire_time_changed(hass, future)
-            await hass.async_block_till_done()
+    # Begin the process of stopping the script (which should stop all runs), and then
+    # let the service calls complete.
+    hass.async_create_task(script_obj.async_stop())
+    finish_service_event.set()
 
-            assert not script_obj.is_running
-            assert script_obj.last_action is None
+    await hass.async_block_till_done()
 
+    assert script_was_runing
+    assert were_no_events
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5164426)